File: packages/sdk-go/security/input_validator.go
// Package security provides security utilities for input validation and risk analysis.
package security

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/ethereum/go-ethereum/common"
)

// ============================================================================
// Validation Result
// ============================================================================

// ValidationResult contains the result of input validation.
type ValidationResult struct {
	// IsValid indicates whether the input is valid.
	IsValid bool `json:"isValid"`

	// Errors contains validation error messages.
	Errors []string `json:"errors"`

	// Warnings contains non-critical warning messages.
	Warnings []string `json:"warnings,omitempty"`

	// NormalizedValue contains the normalized/sanitized value.
	NormalizedValue any `json:"normalizedValue,omitempty"`
}

// ============================================================================
// Validation Options
// ============================================================================

// HexValidationOptions configures hex string validation.
type HexValidationOptions struct {
	// RequirePrefix requires the 0x prefix.
	RequirePrefix bool

	// ExactLength requires an exact length (including 0x prefix).
	ExactLength int

	// MinLength requires a minimum length.
	MinLength int

	// MaxLength requires a maximum length.
	MaxLength int
}

// SanitizeOptions configures string sanitization.
type SanitizeOptions struct {
	// MaxLength is the maximum allowed length.
	MaxLength int

	// EscapeHtml escapes HTML entities.
	EscapeHtml bool
}

// ============================================================================
// Transaction Object
// ============================================================================

// TransactionObject represents a transaction for validation.
type TransactionObject struct {
	From                 string `json:"from,omitempty"`
	To                   string `json:"to,omitempty"`
	Value                string `json:"value,omitempty"`
	Gas                  string `json:"gas,omitempty"`
	GasPrice             string `json:"gasPrice,omitempty"`
	MaxFeePerGas         string `json:"maxFeePerGas,omitempty"`
	MaxPriorityFeePerGas string `json:"maxPriorityFeePerGas,omitempty"`
	Data                 string `json:"data,omitempty"`
	Nonce                string `json:"nonce,omitempty"`
}

// RpcRequestObject represents an RPC request for validation.
type RpcRequestObject struct {
	Jsonrpc string `json:"jsonrpc,omitempty"`
	Method  string `json:"method,omitempty"`
	Params  any    `json:"params,omitempty"`
	Id      any    `json:"id,omitempty"`
}

// ============================================================================
// Input Validator
// ============================================================================

// InputValidator provides comprehensive input validation for wallet security.
type InputValidator struct{}

// NewInputValidator creates a new InputValidator.
func NewInputValidator() *InputValidator {
	return &InputValidator{}
}

// ValidateAddress validates an Ethereum address.
func (v *InputValidator) ValidateAddress(address string) *ValidationResult {
	errors := []string{}
	warnings := []string{}

	// Check for empty
	if address == "" {
		return &ValidationResult{
			IsValid: false,
			Errors:  []string{"Address is required"},
		}
	}

	// Check prefix
	if !strings.HasPrefix(address, "0x") {
		errors = append(errors, "Address must start with 0x")
	}

	// Check length
	if len(address) != 42 {
		errors = append(errors, "Address must be 42 characters")
	}

	// Check for valid hex characters
	if len(address) > 2 {
		hexPart := address[2:]
		matched, _ := regexp.MatchString(`^[0-9a-fA-F]*$`, hexPart)
		if !matched {
			errors = append(errors, "Address contains invalid characters")
		}
	}

	// Validate EIP-55 checksum for mixed case addresses
	if len(errors) == 0 {
		hexPart := address[2:]
		lowerCase := strings.ToLower(hexPart)
		upperCase := strings.ToUpper(hexPart)
		isMixedCase := hexPart != lowerCase && hexPart != upperCase

		if isMixedCase {
			// Verify checksum using go-ethereum's implementation
			checksumAddr := common.HexToAddress(address)
			if checksumAddr.Hex() != address {
				warnings = append(warnings, "Address has mixed case but invalid EIP-55 checksum")
				errors = append(errors, "Address has invalid EIP-55 checksum")
			}
		}
	}

	var normalizedValue any
	if len(errors) == 0 {
		checksumAddr := common.HexToAddress(address)
		normalizedValue = checksumAddr.Hex()
	}

	return &ValidationResult{
		IsValid:         len(errors) == 0,
		Errors:          errors,
		Warnings:        warnings,
		NormalizedValue: normalizedValue,
	}
}

// ValidateHex validates a hex string.
func (v *InputValidator) ValidateHex(hex string, options *HexValidationOptions) *ValidationResult {
	if options == nil {
		options = &HexValidationOptions{RequirePrefix: true}
	}

	errors := []string{}

	// Check for empty
	if hex == "" {
		return &ValidationResult{
			IsValid: false,
			Errors:  []string{"Hex string is required"},
		}
	}

	// Check prefix
	hasPrefix := strings.HasPrefix(hex, "0x")
	if options.RequirePrefix && !hasPrefix {
		errors = append(errors, "Hex string must start with 0x")
	}

	// Get hex content
	hexContent := hex
	if hasPrefix {
		hexContent = hex[2:]
	}
	fullLength := len(hex)
	if !hasPrefix {
		fullLength += 2
	}

	// Check for valid hex characters
	matched, _ := regexp.MatchString(`^[0-9a-fA-F]*$`, hexContent)
	if !matched {
		errors = append(errors, "Invalid hex characters")
	}

	// Check exact length
	if options.ExactLength > 0 && fullLength != options.ExactLength {
		errors = append(errors, "Hex string must be exactly "+strconv.Itoa(options.ExactLength)+" characters")
	}

	// Check minimum length
	if options.MinLength > 0 && fullLength < options.MinLength {
		errors = append(errors, "Hex string must be at least "+strconv.Itoa(options.MinLength)+" characters")
	}

	// Check maximum length
	if options.MaxLength > 0 && fullLength > options.MaxLength {
		errors = append(errors, "Hex string must be at most "+strconv.Itoa(options.MaxLength)+" characters")
	}

	var normalizedValue any
	if len(errors) == 0 {
		if hasPrefix {
			normalizedValue = strings.ToLower(hex)
		} else {
			normalizedValue = strings.ToLower("0x" + hex)
		}
	}

	return &ValidationResult{
		IsValid:         len(errors) == 0,
		Errors:          errors,
		NormalizedValue: normalizedValue,
	}
}

// ValidateChainId validates a chain ID.
func (v *InputValidator) ValidateChainId(chainId any) *ValidationResult {
	errors := []string{}
	var normalizedValue int64

	switch cid := chainId.(type) {
	case string:
		if strings.HasPrefix(cid, "0x") {
			val, err := strconv.ParseInt(cid[2:], 16, 64)
			if err != nil {
				return &ValidationResult{
					IsValid: false,
					Errors:  []string{"Invalid chain ID format"},
				}
			}
			normalizedValue = val
		} else {
			val, err := strconv.ParseInt(cid, 10, 64)
			if err != nil {
				return &ValidationResult{
					IsValid: false,
					Errors:  []string{"Invalid chain ID format"},
				}
			}
			normalizedValue = val
		}
	case int:
		normalizedValue = int64(cid)
	case int64:
		normalizedValue = cid
	case uint64:
		normalizedValue = int64(cid)
	case float64:
		normalizedValue = int64(cid)
	default:
		return &ValidationResult{
			IsValid: false,
			Errors:  []string{"Chain ID must be a number or string"},
		}
	}

	// Check if positive
	if normalizedValue <= 0 {
		errors = append(errors, "Chain ID must be positive")
	}

	return &ValidationResult{
		IsValid:         len(errors) == 0,
		Errors:          errors,
		NormalizedValue: normalizedValue,
	}
}

// ValidateTransaction validates a transaction object.
func (v *InputValidator) ValidateTransaction(tx *TransactionObject) *ValidationResult {
	errors := []string{}
	warnings := []string{}

	// Check for from address (required)
	if tx.From == "" {
		errors = append(errors, "Transaction must have a from address")
	} else {
		fromResult := v.ValidateAddress(tx.From)
		if !fromResult.IsValid {
			errors = append(errors, "Invalid from address: "+strings.Join(fromResult.Errors, ", "))
		}
	}

	// Check to address (optional for contract deployment)
	if tx.To != "" {
		toResult := v.ValidateAddress(tx.To)
		if !toResult.IsValid {
			errors = append(errors, "Invalid to address: "+strings.Join(toResult.Errors, ", "))
		}
	} else if tx.Data == "" || tx.Data == "0x" {
		warnings = append(warnings, "Transaction has no to address and no data - this may fail")
	}

	// Validate value if present
	if tx.Value != "" {
		valueResult := v.ValidateHex(tx.Value, nil)
		if !valueResult.IsValid {
			errors = append(errors, "Invalid value: "+strings.Join(valueResult.Errors, ", "))
		}
	}

	// Validate gas if present
	if tx.Gas != "" {
		gasResult := v.ValidateHex(tx.Gas, nil)
		if !gasResult.IsValid {
			errors = append(errors, "Invalid gas: "+strings.Join(gasResult.Errors, ", "))
		}
	}

	// Validate gasPrice if present
	if tx.GasPrice != "" {
		gasPriceResult := v.ValidateHex(tx.GasPrice, nil)
		if !gasPriceResult.IsValid {
			errors = append(errors, "Invalid gasPrice: "+strings.Join(gasPriceResult.Errors, ", "))
		}
	}

	// Validate maxFeePerGas if present
	if tx.MaxFeePerGas != "" {
		maxFeeResult := v.ValidateHex(tx.MaxFeePerGas, nil)
		if !maxFeeResult.IsValid {
			errors = append(errors, "Invalid maxFeePerGas: "+strings.Join(maxFeeResult.Errors, ", "))
		}
	}

	// Validate maxPriorityFeePerGas if present
	if tx.MaxPriorityFeePerGas != "" {
		maxPriorityResult := v.ValidateHex(tx.MaxPriorityFeePerGas, nil)
		if !maxPriorityResult.IsValid {
			errors = append(errors, "Invalid maxPriorityFeePerGas: "+strings.Join(maxPriorityResult.Errors, ", "))
		}
	}

	// Validate data if present
	if tx.Data != "" && tx.Data != "0x" {
		dataResult := v.ValidateHex(tx.Data, nil)
		if !dataResult.IsValid {
			errors = append(errors, "Invalid data: "+strings.Join(dataResult.Errors, ", "))
		}
	}

	// Validate nonce if present
	if tx.Nonce != "" {
		nonceResult := v.ValidateHex(tx.Nonce, nil)
		if !nonceResult.IsValid {
			errors = append(errors, "Invalid nonce: "+strings.Join(nonceResult.Errors, ", "))
		}
	}

	return &ValidationResult{
		IsValid:  len(errors) == 0,
		Errors:   errors,
		Warnings: warnings,
	}
}

// ValidateRpcRequest validates an RPC request object.
func (v *InputValidator) ValidateRpcRequest(request *RpcRequestObject) *ValidationResult {
	errors := []string{}

	// Check for method (required)
	if request.Method == "" {
		errors = append(errors, "RPC request must have a method")
	}

	// Check params if present
	if request.Params != nil {
		if _, ok := request.Params.([]any); !ok {
			if _, ok := request.Params.([]interface{}); !ok {
				errors = append(errors, "Params must be an array")
			}
		}
	}

	// Check id if present
	if request.Id != nil {
		switch request.Id.(type) {
		case int, int64, float64, string:
			// Valid types
		default:
			errors = append(errors, "ID must be a number or string")
		}
	}

	return &ValidationResult{
		IsValid: len(errors) == 0,
		Errors:  errors,
	}
}

// SanitizeString sanitizes string input.
func (v *InputValidator) SanitizeString(input string, options *SanitizeOptions) string {
	if options == nil {
		options = &SanitizeOptions{MaxLength: 10000, EscapeHtml: true}
	}

	result := input

	// Remove null bytes
	result = strings.ReplaceAll(result, "\x00", "")

	// Trim whitespace
	result = strings.TrimSpace(result)

	// Limit length
	maxLength := options.MaxLength
	if maxLength <= 0 {
		maxLength = 10000
	}
	if len(result) > maxLength {
		result = result[:maxLength]
	}

	// Escape HTML if enabled
	if options.EscapeHtml {
		result = strings.ReplaceAll(result, "&", "&amp;")
		result = strings.ReplaceAll(result, "<", "&lt;")
		result = strings.ReplaceAll(result, ">", "&gt;")
		result = strings.ReplaceAll(result, "\"", "&quot;")
		result = strings.ReplaceAll(result, "'", "&#039;")
	}

	return result
}

// ============================================================================
// Utility Functions
// ============================================================================

// Default validator instance
var defaultValidator = NewInputValidator()

// IsValidAddress checks if an address is valid.
func IsValidAddress(address string) bool {
	return defaultValidator.ValidateAddress(address).IsValid
}

// IsValidHex checks if a hex string is valid.
func IsValidHex(hex string, options *HexValidationOptions) bool {
	return defaultValidator.ValidateHex(hex, options).IsValid
}

// IsValidChainId checks if a chain ID is valid.
func IsValidChainId(chainId any) bool {
	return defaultValidator.ValidateChainId(chainId).IsValid
}

// IsValidTransactionObject checks if a transaction object is valid.
func IsValidTransactionObject(tx *TransactionObject) bool {
	return defaultValidator.ValidateTransaction(tx).IsValid
}

// IsValidRpcRequest checks if an RPC request is valid.
func IsValidRpcRequest(request *RpcRequestObject) bool {
	return defaultValidator.ValidateRpcRequest(request).IsValid
}

// SanitizeString sanitizes a string using the default validator.
func SanitizeString(input string, options *SanitizeOptions) string {
	return defaultValidator.SanitizeString(input, options)
}

File: packages/sdk-ts/core/src/security/inputValidator.ts
/**
 * Input Validator
 * Comprehensive input validation for wallet security
 */

import { getAddress, isAddress } from 'viem'

/**
 * Validation result interface
 */
export interface ValidationResult {
  isValid: boolean
  errors: string[]
  warnings?: string[]
  normalizedValue?: unknown
}

/**
 * Hex validation options
 */
export interface HexValidationOptions {
  requirePrefix?: boolean
  exactLength?: number
  minLength?: number
  maxLength?: number
}

/**
 * String sanitization options
 */
export interface SanitizeOptions {
  maxLength?: number
  escapeHtml?: boolean
}

/**
 * Transaction object interface
 */
export interface TransactionObject {
  from?: string
  to?: string
  value?: string
  gas?: string
  gasPrice?: string
  maxFeePerGas?: string
  maxPriorityFeePerGas?: string
  data?: string
  nonce?: string
}

/**
 * RPC request interface
 */
export interface RpcRequestObject {
  jsonrpc?: string
  method?: unknown
  params?: unknown
  id?: unknown
}

/**
 * EIP-55 checksum address conversion using viem's keccak256-based implementation
 * Returns the checksummed address or null if invalid
 */
function toChecksumAddress(address: string): string | null {
  try {
    return getAddress(address)
  } catch {
    return null
  }
}

/**
 * Check if address has valid EIP-55 checksum using viem's strict validation
 */
function hasValidChecksum(address: string): boolean {
  // All lowercase or all uppercase is valid (no checksum to verify)
  const hexPart = address.slice(2)
  const lowerCase = hexPart.toLowerCase()
  const upperCase = hexPart.toUpperCase()

  if (hexPart === lowerCase || hexPart === upperCase) {
    return true
  }

  // For mixed case, verify checksum using viem's strict mode
  // isAddress with strict: true validates EIP-55 checksum
  return isAddress(address, { strict: true })
}

/**
 * Input Validator class
 */
export class InputValidator {
  /**
   * Validate Ethereum address
   */
  validateAddress(address: string): ValidationResult {
    const errors: string[] = []
    const warnings: string[] = []

    // Check for null/undefined
    if (!address || typeof address !== 'string') {
      return {
        isValid: false,
        errors: ['Address is required and must be a string'],
      }
    }

    // Check prefix
    if (!address.startsWith('0x')) {
      errors.push('Address must start with 0x')
    }

    // Check length
    if (address.length !== 42) {
      errors.push('Address must be 42 characters')
    }

    // Check for valid hex characters
    const hexPart = address.slice(2)
    if (!/^[0-9a-fA-F]*$/.test(hexPart)) {
      errors.push('Address contains invalid characters')
    }

    // Validate EIP-55 checksum for mixed case addresses
    if (errors.length === 0) {
      const hexPart = address.slice(2)
      const lowerCase = hexPart.toLowerCase()
      const upperCase = hexPart.toUpperCase()
      const isMixedCase = hexPart !== lowerCase && hexPart !== upperCase

      if (!hasValidChecksum(address)) {
        // Mixed case with invalid checksum - add warning
        if (isMixedCase) {
          warnings.push('Address has mixed case but invalid EIP-55 checksum')
        }
        errors.push('Address has invalid EIP-55 checksum')
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings: warnings.length > 0 ? warnings : undefined,
      // Return EIP-55 checksummed address as normalized value
      normalizedValue: errors.length === 0 ? toChecksumAddress(address) : undefined,
    }
  }

  /**
   * Validate hex string
   */
  validateHex(hex: string, options: HexValidationOptions = {}): ValidationResult {
    const errors: string[] = []
    const { requirePrefix = true, exactLength, minLength, maxLength } = options

    // Check for null/undefined
    if (!hex || typeof hex !== 'string') {
      return {
        isValid: false,
        errors: ['Hex string is required'],
      }
    }

    // Check prefix
    const hasPrefix = hex.startsWith('0x')
    if (requirePrefix && !hasPrefix) {
      errors.push('Hex string must start with 0x')
    }

    // Get hex content
    const hexContent = hasPrefix ? hex.slice(2) : hex
    const fullLength = hasPrefix ? hex.length : hex.length + 2

    // Check for valid hex characters
    if (!/^[0-9a-fA-F]*$/.test(hexContent)) {
      errors.push('Invalid hex characters')
    }

    // Check exact length
    if (exactLength !== undefined && fullLength !== exactLength) {
      errors.push(`Hex string must be exactly ${exactLength} characters`)
    }

    // Check minimum length
    if (minLength !== undefined && fullLength < minLength) {
      errors.push(`Hex string must be at least ${minLength} characters`)
    }

    // Check maximum length
    if (maxLength !== undefined && fullLength > maxLength) {
      errors.push(`Hex string must be at most ${maxLength} characters`)
    }

    return {
      isValid: errors.length === 0,
      errors,
      normalizedValue:
        errors.length === 0 ? (hasPrefix ? hex : `0x${hex}`).toLowerCase() : undefined,
    }
  }

  /**
   * Validate chain ID
   */
  validateChainId(chainId: unknown): ValidationResult {
    const errors: string[] = []
    let normalizedValue: number | undefined

    // Handle string input
    if (typeof chainId === 'string') {
      // Try hex format
      if (chainId.startsWith('0x')) {
        normalizedValue = Number.parseInt(chainId, 16)
      } else {
        normalizedValue = Number.parseInt(chainId, 10)
      }
    } else if (typeof chainId === 'number') {
      normalizedValue = chainId
    } else {
      return {
        isValid: false,
        errors: ['Chain ID must be a number or string'],
      }
    }

    // Check if parsing succeeded
    if (Number.isNaN(normalizedValue)) {
      return {
        isValid: false,
        errors: ['Invalid chain ID format'],
      }
    }

    // Check if positive
    if (normalizedValue <= 0) {
      errors.push('Chain ID must be positive')
    }

    // Check if integer
    if (!Number.isInteger(normalizedValue)) {
      errors.push('Chain ID must be an integer')
    }

    return {
      isValid: errors.length === 0,
      errors,
      normalizedValue: errors.length === 0 ? normalizedValue : undefined,
    }
  }

  /**
   * Validate transaction object
   */
  validateTransaction(tx: TransactionObject): ValidationResult {
    const errors: string[] = []
    const warnings: string[] = []

    // Check for from address (required)
    if (!tx.from) {
      errors.push('Transaction must have a from address')
    } else {
      const fromResult = this.validateAddress(tx.from)
      if (!fromResult.isValid) {
        errors.push(`Invalid from address: ${fromResult.errors.join(', ')}`)
      }
    }

    // Check to address (optional for contract deployment)
    if (tx.to) {
      const toResult = this.validateAddress(tx.to)
      if (!toResult.isValid) {
        errors.push(`Invalid to address: ${toResult.errors.join(', ')}`)
      }
    } else if (!tx.data || tx.data === '0x') {
      warnings.push('Transaction has no to address and no data - this may fail')
    }

    // Validate value if present
    if (tx.value !== undefined) {
      const valueResult = this.validateHex(tx.value)
      if (!valueResult.isValid) {
        errors.push(`Invalid value: ${valueResult.errors.join(', ')}`)
      }
    }

    // Validate gas if present
    if (tx.gas !== undefined) {
      const gasResult = this.validateHex(tx.gas)
      if (!gasResult.isValid) {
        errors.push(`Invalid gas: ${gasResult.errors.join(', ')}`)
      }
    }

    // Validate gasPrice if present
    if (tx.gasPrice !== undefined) {
      const gasPriceResult = this.validateHex(tx.gasPrice)
      if (!gasPriceResult.isValid) {
        errors.push(`Invalid gasPrice: ${gasPriceResult.errors.join(', ')}`)
      }
    }

    // Validate maxFeePerGas if present
    if (tx.maxFeePerGas !== undefined) {
      const maxFeeResult = this.validateHex(tx.maxFeePerGas)
      if (!maxFeeResult.isValid) {
        errors.push(`Invalid maxFeePerGas: ${maxFeeResult.errors.join(', ')}`)
      }
    }

    // Validate maxPriorityFeePerGas if present
    if (tx.maxPriorityFeePerGas !== undefined) {
      const maxPriorityResult = this.validateHex(tx.maxPriorityFeePerGas)
      if (!maxPriorityResult.isValid) {
        errors.push(`Invalid maxPriorityFeePerGas: ${maxPriorityResult.errors.join(', ')}`)
      }
    }

    // Validate data if present
    if (tx.data !== undefined && tx.data !== '0x') {
      const dataResult = this.validateHex(tx.data)
      if (!dataResult.isValid) {
        errors.push(`Invalid data: ${dataResult.errors.join(', ')}`)
      }
    }

    // Validate nonce if present
    if (tx.nonce !== undefined) {
      const nonceResult = this.validateHex(tx.nonce)
      if (!nonceResult.isValid) {
        errors.push(`Invalid nonce: ${nonceResult.errors.join(', ')}`)
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings: warnings.length > 0 ? warnings : undefined,
    }
  }

  /**
   * Validate RPC request object
   */
  validateRpcRequest(request: RpcRequestObject): ValidationResult {
    const errors: string[] = []

    // Check for method (required)
    if (!request.method) {
      errors.push('RPC request must have a method')
    } else if (typeof request.method !== 'string') {
      errors.push('Method must be a string')
    }

    // Check params if present
    if (request.params !== undefined && !Array.isArray(request.params)) {
      errors.push('Params must be an array')
    }

    // Check id if present
    if (request.id !== undefined) {
      if (typeof request.id !== 'number' && typeof request.id !== 'string') {
        errors.push('ID must be a number or string')
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
    }
  }

  /**
   * Sanitize string input
   */
  sanitizeString(input: string, options: SanitizeOptions = {}): string {
    const { maxLength = 10000, escapeHtml = true } = options

    if (typeof input !== 'string') {
      return ''
    }

    let result = input

    // Remove null bytes
    // biome-ignore lint/suspicious/noControlCharactersInRegex: intentionally removing null bytes for security
    result = result.replace(/\x00/g, '')

    // Trim whitespace
    result = result.trim()

    // Limit length
    if (result.length > maxLength) {
      result = result.slice(0, maxLength)
    }

    // Escape HTML if enabled
    if (escapeHtml) {
      result = result
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#039;')
    }

    return result
  }
}

/**
 * Create a new InputValidator instance
 */
export function createInputValidator(): InputValidator {
  return new InputValidator()
}

// =============================================================================
// Utility Functions
// =============================================================================

const defaultValidator = new InputValidator()

/**
 * Check if address is valid
 */
export function isValidAddress(address: string): boolean {
  return defaultValidator.validateAddress(address).isValid
}

/**
 * Check if hex string is valid
 */
export function isValidHex(hex: string, options?: HexValidationOptions): boolean {
  return defaultValidator.validateHex(hex, options).isValid
}

/**
 * Check if chain ID is valid
 */
export function isValidChainId(chainId: unknown): boolean {
  return defaultValidator.validateChainId(chainId).isValid
}

/**
 * Check if transaction object is valid
 */
export function isValidTransactionObject(tx: TransactionObject): boolean {
  return defaultValidator.validateTransaction(tx).isValid
}

/**
 * Check if RPC request is valid
 */
export function isValidRpcRequest(request: RpcRequestObject): boolean {
  return defaultValidator.validateRpcRequest(request).isValid
}

/**
 * Sanitize string
 */
export function sanitizeString(input: string, options?: SanitizeOptions): string {
  return defaultValidator.sanitizeString(input, options)
}
